pyanaconda/kickstart.py

```diff
@@ -72,63 +69,19 @@ def check_kickstart_error():
 
 
 class AnacondaKSScript(KSScript):
-    """ Execute a kickstart script
-
-        This will write the script to a file named /tmp/ks-script- before
-        execution.
-        Output is logged by the program logger, the path specified by --log
-        or to /tmp/ks-script-\\*.log
-    """
     def run(self, chroot):
-        """ Run the kickstart script
-            @param chroot directory path to chroot into before execution
-        """
-        if self.inChroot:
-            scriptRoot = chroot
-        else:
-            scriptRoot = "/"
-
-        (fd, path) = tempfile.mkstemp("", "ks-script-", scriptRoot + "/tmp")
-
-        os.write(fd, self.script.encode("utf-8"))
-        os.close(fd)
-        os.chmod(path, 0o700)
-
-        # Always log stdout/stderr from scripts.  Using --log just lets you
-        # pick where it goes.  The script will also be logged to program.log
-        # because of execWithRedirect.
-        if self.logfile:
-            if self.inChroot:
-                messages = "%s/%s" % (scriptRoot, self.logfile)
-            else:
-                messages = self.logfile
-
-            d = os.path.dirname(messages)
-            if not os.path.exists(d):
-                os.makedirs(d)
-        else:
-            # Always log outside the chroot, we copy those logs into the
-            # chroot later.
-            messages = "/tmp/%s.log" % os.path.basename(path)
-
-        with open_with_perm(messages, "w", 0o600) as fp:
-            rc = util.execWithRedirect(self.interp, ["/tmp/%s" % os.path.basename(path)],
-                                       stdout=fp,
-                                       root=scriptRoot)
-
-        if rc != 0:
-            script_log.error("Error code %s running the kickstart script at line %s", rc, self.lineno)
-            if self.errorOnFail:
-                err = ""
-                with open(messages, "r") as fp:
-                    err = "".join(fp.readlines())
-
-                # Show error dialog even for non-interactive
-                flags.ksprompt = True
-
-                errorHandler.cb(ScriptError(self.lineno, err))
-                util.ipmi_report(IPMI_ABORTED)
-                sys.exit(0)
+        rc, log_file = run_script(self, chroot)
```

Comment:
Another challenge: you need to pass handle the result in anaconda process so that the `--erroronfail` option of the scripts can be still supported.

Reply:
So I tested this option and it works exactly the same in master and in this PR

Reply:
Hm, I need to check again, maybe I was too fast here.

Reply:
Ok, so for example I am adding this pre script to the container kickstart test (I believe you can reproduce it anywhere, using kstest just for convinience):
```
%pre-install --erroronfail
echo stdout output
>&2 echo stderr_output
exit 1
%end
```
On rawhide it fails with this Dialog:

![Screenshot from 2024-08-05 16-23-05](https://github.com/user-attachments/assets/a5a11dc2-0c75-4df3-994c-8b3c8e844772)


With the patch, it ends with 
```
14:39:58,651 WARNING org.fedoraproject.Anaconda.Modules.Runtime:INFO:anaconda.core.threads:Thread Done: AnaTaskThread-RunScriptsTask-1 (140160954009280)
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:WARNING:dasbus.server.handler:The call org.fedoraproject.Anaconda.Task.Finish has failed with an exception:
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:Traceback (most recent call last):
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib/python3.13/site-packages/dasbus/server/handler.py", line 455, in _method_callback
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    result = self._handle_call(
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:        interface_name,
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ...<2 lines>...
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:        **additional_args
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    )
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib/python3.13/site-packages/dasbus/server/handler.py", line 265, in _handle_call
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    return handler(*parameters, **additional_args)
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task_interface.py", line 114, in Finish
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    self.implementation.finish()
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 173, in finish
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    thread_manager.raise_if_error(self._thread_name)
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 171, in raise_if_error
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    raise exc_info[1]
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 280, in run
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    threading.Thread.run(self)
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~~~~~~~~~~~^^^^^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/threading.py", line 990, in run
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    self._target(*self._args, **self._kwargs)
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 94, in _thread_run_callback
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    self._task_run_callback()
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~~~~~~~~~~~~~~^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 107, in _task_run_callback
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    self._set_result(self.run())
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:                     ~~~~~~~~^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/runtime/scripts/runtime.py", line 51, in run
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    script.run("/")
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~^^^^^
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/kickstart.py", line 82, in run
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    errorHandler.cb(ScriptError(self.lineno, err))
14:39:58,750 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
14:39:58,758 WARNING org.fedoraproject.Anaconda.Modules.Runtime:  File "/usr/lib64/python3.13/site-packages/pyanaconda/errors.py", line 226, in cb
14:39:58,758 WARNING org.fedoraproject.Anaconda.Modules.Runtime:    raise exn
14:39:58,758 WARNING org.fedoraproject.Anaconda.Modules.Runtime:pyanaconda.errors.ScriptError
```